### PR TITLE
Osdag Workshop Certificate

### PR DESCRIPTION
**Pull request by binson-b**

### CODE REVIEW

File: certificate/models.py [hunk 1]
```diff
@@ -163,6 +163,9 @@ class Esim_faculty(models.Model):
 class Osdag_WS(models.Model):
     name = models.CharField(max_length=200)
     email = models.EmailField()
+    college = models.CharField(max_length=200, null=True, blank=True)
+    ws_date = models.CharField(max_length = 100, null=True, blank=True)
+    ws_year = models.CharField(max_length = 100, null=True, blank=True)
```
Comment (prathamesh920) [inline certificate/models.py]: Keep this ```date``` anad ```year``` only, that's better.
No need of ```ws_```
So ```Osdag_WS.date```  is better than ```Osdag_WS.ws_date```
Even the class name is not good but for now let it be.

File: certificate/views.py [hunk 2]
```diff
@@ -143,7 +143,7 @@ def verification(serial, _type):
                 elif purpose == 'Osdag Workshop':
                     faculty = Osdag_WS.objects.get(email=certificate.email)
                     detail = OrderedDict([('Name', name), ('Event', purpose),
-                        ('Days', '4 June'), ('Year', year)])
+                        ('Days', faculty.ws_date), ('Year', year)])
```
Comment (prathamesh920) [inline certificate/views.py]: If possible then change ```faculty``` to ```osdag_workshop```

File: certificate/views.py [hunk 3]
```diff
@@ -1135,17 +1127,22 @@ def osdag_workshop_download(request):
                 user = user[0]
         name = user.name
         purpose = user.purpose
-        year = '16'
+        year = user.ws_year[2:]
```
Comment (prathamesh920) [inline certificate/views.py]: Even here the ```user``` is not suitable.
But let it be as it was from start and the model structure and attribute names are such that at this point its difficult to change.

File: certificate/views.py [hunk 4]
```diff
@@ -1135,17 +1127,22 @@ def osdag_workshop_download(request):
...
         id =  int(user.id)
         hexa = hex(id).replace('0x','').zfill(6).upper()
         serial_no = '{0}{1}{2}{3}'.format(purpose, year, hexa, type)
         serial_key = (hashlib.sha1(serial_no)).hexdigest()
         file_name = '{0}{1}'.format(email,id)
         file_name = file_name.replace('.', '')
+        details = {
+        'name': name, 'year': user.ws_year,
```
Comment (prathamesh920) [inline certificate/views.py]: I think here if you add indentation then will be better.

**Review by prathamesh920 — CHANGES_REQUESTED** [review]
Few comments.
Else LGTM

**Comment by prathamesh920** [comment]
Also did you check for same user but different years or an old existing user with no years!

File: certificate/views.py (hunk 3 above)
Comment (binson-b) [reply certificate/views.py]: ok. I kept it as it is

File: certificate/views.py [hunk 5]
```diff
@@ -1124,7 +1124,12 @@ def osdag_workshop_download(request):
                 return render_to_response('osdag_workshop_download.html',
                         context, context_instance=ci)
             else:
-                user = user[0]
+                if len(user) > 1:
```
Comment (prathamesh920) [inline certificate/views.py]: Change the variable name ```user``` to ```users```

File: certificate/views.py [hunk 6]
```diff
@@ -1124,7 +1124,12 @@ def osdag_workshop_download(request):
...
+                    user_list = [(each_user, int(each_user.year)) for each_user in user]
```
Comment (prathamesh920) [inline certificate/views.py]: ```for user in users"``` is better than ```each user```

File: certificate/views.py [hunk 7]
```diff
@@ -1124,7 +1124,12 @@ def osdag_workshop_download(request):
...
+                    sorted(user_list, key=lambda x: x[1], reverse=True)
```
Comment (prathamesh920) [inline certificate/views.py]: Instead of this complex sorting, why can't we have django filter ordered by year and give the latest user.

**Review by prathamesh920 — CHANGES_REQUESTED** [review]
Kindly resolve the comments and comment after you resolve.
Also, selecting the latest user as per the year can be logically incorrect.
If I want a certificate for the previous year then how can I get it.
Kindly think on this.

File: certificate/models.py (hunk 1 above)
Comment (binson-b) [reply certificate/models.py]: Done

File: certificate/views.py (hunk 2 above)
Comment (binson-b) [reply certificate/views.py]: Done

File: certificate/views.py (hunk 4 above)
Comment (binson-b) [reply certificate/views.py]: done

File: certificate/views.py (hunk 5 above)
Comment (binson-b) [reply certificate/views.py]: ok i will do it.

File: certificate/views.py (hunk 6 above)
Comment (binson-b) [reply certificate/views.py]: ok,

File: certificate/views.py (hunk 7 above)
Comment (binson-b) [reply certificate/views.py]: removed it. added start and end date field

**Comment by binson-b** [comment]
@prathamesh920 i have removed the complex code and gave an option to select the date of the workshop, so that all year certificate are avaliable to download.

**Comment by prathamesh920** [comment]
I think currently only 2018 certificates are available for osdag and no other! 
So for users its an extra work to select the only value 2018. This seems unnecessary at this point. 
Also the  date comparison query can be simplified further!
If this is very urgent and works well then create an issue and then i will merge it.
Else resolve the issue in this PR itself.  Mainly the date selection process!

**Comment by binson-b** [comment]
@prathamesh920 it works well,  lets keep the date because it will help in filtering out to one user. 
and about date comparison, i am just converting the date to DateTime obj and using it  in the filter

**Comment by prathamesh920** [comment]
Merging for now

**Pull request merged**